src/RA_Defs.cpp

```diff
@@ -9,55 +9,74 @@ GetParseErrorFunc GetJSONParseErrorStr = GetParseError_En;
 
 #endif
 
-static_assert(sizeof(BYTE*) == sizeof(char*), "dangerous cast ahead");
-char* DataStreamAsString(DataStream& stream)
-{
-    return reinterpret_cast<char*>(stream.data());
-}
+namespace ra {
 
-std::string Narrow(const wchar_t* wstr)
+std::string Narrow(const std::wstring& wstr)
 {
-    static std::wstring_convert< std::codecvt_utf8_utf16< wchar_t >, wchar_t > converter;
-    return converter.to_bytes(wstr);
+    return Narrow(wstr.c_str());
 }
 
-std::string Narrow(const std::wstring& wstr)
+std::string Narrow(std::wstring&& wstr) noexcept
 {
-    static std::wstring_convert< std::codecvt_utf8_utf16< wchar_t >, wchar_t > converter;
-    return converter.to_bytes(wstr);
+    auto wwstr{ std::move_if_noexcept(wstr) };
+    return Narrow(wwstr);
 }
 
-std::wstring Widen(const char* str)
+
+std::string Narrow(const wchar_t* wstr)
 {
-    static std::wstring_convert< std::codecvt_utf8_utf16< wchar_t >, wchar_t > converter;
-    return converter.from_bytes(str);
+    auto state{ std::mbstate_t{} };
+    auto len{ 1 + std::wcsrtombs(nullptr, &wstr, 0_z, &state) };
+
+    // doesn't work with unique_ptr, though it works with sprintf and the functions in RAWeb
+    std::string str;
+    str.reserve(len);
```

Comment:
could have sworn I removed that comment. Yeah I tried using the array of unique_ptr before using a string but it didn't work. I don't like static arrays and I don't like using new directly.